compiler/src/dotty/tools/dotc/Compiler.scala

```diff
@@ -74,9 +74,9 @@ class Compiler {
          new HoistSuperArgs,         // Hoist complex arguments of supercalls to enclosing scope
          new SpecializeApplyMethods, // Adds specialized methods to FunctionN
          new RefChecks) ::           // Various checks mostly related to abstract members and overriding
+    List(new TryCatchPatterns,       // Compile cases in try/catch
+         new PatternMatcher) ::      // Compile pattern matches
```

Comment:
Adding one extra group of phase isn't great for performance (bcause it means one extra tree traversal), so it'd be good to figure out whether there really is no way around that.

Reply:
Happy to try some different ordering/groupings, if y'all reviewers have any to suggest? I don't have much context / bigger-picture knowledge to bring to this.

Reply:
I don't think a different ordering will save us here, some logic change in PatternMatcher might be needed instead but I don't know enough about the interactions between ExplicitOuter and PatternMatcher to say. I'll note that if I move the class definition after the pattern match, then everything works fine:
```scala
class C1 {
  new C2 match {
    case c: C2 =>
  }
  private class C2
}

object Test extends App {
  new C1
}
```
So perhaps we could get away with just reordering statements so that class definitions come after all other definitions in a class body. We already do some reordering in FirstTransform so this isn't too far-fetched: https://github.com/lampepfl/dotty/blob/ec15557c1bcce768c1cc8c59b7bc4cb6e8d5e9dc/compiler/src/dotty/tools/dotc/transform/FirstTransform.scala#L66-L67

Reply:
I don't think reordering will work. Consider this example:

```scala
class C1 {
  private class C3 {
    new C2 match {
      case c: C2 =>
    }
  }
  private class C2
  new C3
}
```

If `C3` precedes `C2`, it succeeds. But if `C2` comes first, we get the same bug.

Reply:
Hmmph, indeed, so I don't have a good idea of what to do here. You could try moving PatternMatcher to the previous group, but I doubt that would just work.

Reply:
> You could try moving PatternMatcher to the previous group

Dale has pushed a commit that does that (`ElimRepeated` also needs to move); let's see what CI thinks.

Reply:
Moving two things at once seems pretty daring :). PatternMatcher only checks that it runsAfter ElimRepeated, is that incorrect and actually needs to be runsAfterGroupOf for some reason?

Reply:
> Moving two things at once seems pretty daring

That's how Dale and I roll 😎